lib/charms/vault_k8s/v0/vault_tls.py

```diff
@@ -0,0 +1,451 @@
+            raise PeerSecretError(secret_name=CA_CERTIFICATE_JUJU_SECRET_KEY)
+
+    def _set_ca_certificate_secret_in_peer_relation(
+        self,
+        private_key: str,
+        certificate: str,
+    ) -> None:
+        """Set the vault CA certificate secret in the peer relation.
+
+        Args:
+            private_key: Private key
+            certificate: certificate
+        """
+        if not self.charm._is_peer_relation_created():
+            raise RuntimeError("Peer relation not created")
+        juju_secret_content = {
+            "privatekey": private_key,
+            "certificate": certificate,
+        }
+        juju_secret = self.charm.app.add_secret(
+            juju_secret_content, label=CA_CERTIFICATE_JUJU_SECRET_LABEL
+        )
+        peer_relation = self.charm.model.get_relation(self.peer_relation)
+        peer_relation.data[self.charm.app].update({CA_CERTIFICATE_JUJU_SECRET_KEY: juju_secret.id})
+        logger.debug("Vault CA certificate secret set in peer relation")
+
+    def _remove_ca_certificate_secret_in_peer_relation(self) -> None:
+        """Removes the vault CA certificate secret in the peer relation.
+
+        Only the key from the relation is removed. The actual juju secret is unmodified.
+        """
+        if not self.charm._is_peer_relation_created():
+            raise RuntimeError("Peer relation not created")
+        peer_relation = self.model.get_relation(self.peer_relation)
+        if peer_relation:
+            peer_relation.data[self.charm.app].pop(CA_CERTIFICATE_JUJU_SECRET_KEY)
+            logger.debug("Vault CA certificate secret removed from peer relation")
+
+    def _ca_certificate_set_in_peer_relation(self) -> bool:
+        """Returns whether CA certificate is stored in peer relation data."""
+        try:
+            ca_private_key, ca_certificate = self._get_ca_certificate_secret_in_peer_relation()
+            if ca_private_key and ca_certificate:
+                return True
+        except PeerSecretError:
+            return False
+        return False
+
+    """TODO: The following are methods that go through the container.
+    They will be extracted out in the future."""
+
+    def _remove_all_certs_from_workload(self) -> None:
+        """Removes the certificate files that are used for authentication."""
+        self._remove_tls_file_from_workload(File.CA)
+        self._remove_tls_file_from_workload(File.CERT)
+        self._remove_tls_file_from_workload(File.CSR)
+        logger.debug("Removed existing certificate files from workload.")
+
+    def _reload_vault_container(self) -> None:
+        """Sends a SIGHUP signal to the container which reloads Vault's files. Fails gracefully."""
+        try:
+            self.charm._container.send_signal(SIGHUP, self.charm._container_name)
+            logger.debug("Container restart requested")
+        except APIError:
+            logger.debug("Couldn't send signal to container. Proceeding normally.")
```

Comment:
In the future, we likely will want to call back to the charm code as both the machine and k8s versions of the charm will implement this differently. We'll also want to get rid of references to pebble in this file.

The same applies at a couple of places in this file.

I'm ok with keeping it as is for the moment though and not make this change right now. But this is worth keeping in mind.

Reply:
All of the container methods in this file are grouped at the bottom, and will be easy to replace once we decide what method we want to follow for sharing this code. Currently, my main two ideas are:

1) passing in a variable called substrate (which is defined at the top of this file) which we use to switch methods or
2) passing in a container/machine object that share an abstract base class and virtual methods for pull, push, delete, and reload (as of now, these are the ones that are needed.)